Add shell provisioner

ellmetha opened the pull request:
Depends on #54. :wink:

hsoft commented:
We're again victims of flaky tests. See https://github.com/lxdock/lxdock/pull/55#issuecomment-290278050 . We have to find some solution to this.

hsoft commented on lxdock/provisioners/base.py:
```diff
@@ -91,6 +92,6 @@ def provision(self):
-        return os.path.join(self.homedir, relative_path)
+    def homedir_expanded_path(self, path):
+        """ Expands the considered path with the path of the home homedir if applicable. """
+        return os.path.join(self.homedir, path) if not os.path.isabs(path) else path
```
In your initial commit, I wanted to suggest that you do that, but it turns out that it's not needed: https://docs.python.org/3/library/os.path.html#os.path.join:

`If a component is an absolute path, all previous components are thrown away and joining continues from the absolute path component.`

So I said nothing.

ellmetha replied:
Fixed! :wink:

hsoft commented:
Isn't the "steps" concept redundant? We can already organize our provisioning items in ordered steps. What do we gain by having sub-steps?

hsoft commented:
I also notice that piping (`|`, `>`, `<`) doesn't work for inline shell. I'm not quite sure about the difficulty involved in supporting that, but we should at least mention the limitation in the doc.

ellmetha commented:
@hsoft you're right, ``steps`` are not necessary. I've changed this behaviour. I've also added a note related to the use of common shell patterns (like `>`/`|`/...) to the documentation. It turns out that `lxc exec` doesn't use a shell internally so the problem can be overcame by specifying commands such as `sh -c 'ls -l > /tmp/test`.
